[stdlib] Set, Dictionary: Hash table unification, general cleanup before ABI stability

lorentey opened the pull request:
(Resubmitted from #18790, which was getting unwieldy.)

rdar://problem/28187150

This huge PR refactors `Set` and `Dictionary` to simplify internal structure and to unify low-level hash table operations between the two data structures. It also paves the way for optimizations in upcoming PRs.

- Use a new `_HashTable` struct to unify low-level hashing operations across Set and Dictionary.
- Store the capacity directly inside the storage header. This allows the maximum load factor to be controlled by non-inlinable code, and improves insertion performance.
- Reorganize storage layout. Storage instances are now laid out like this:

    | Field | Description |
    |---|---|
    | `_count` | Count of occupied entries in the table |
    | `_capacity` | Maximum count supported without resizing. (NEW) |
    | `_scale` | The size class of the table. Number of buckets is `1 << _scale`. (NEW) |
    | `_seed` | The hash seed for this instance. |
    | `_rawKeys` | Pointer to the start of tail-allocated region used for storing keys. |
    | `_rawValues` | Pointer to the start of tail-allocated region used for storing values. |
    | `<BITMAP>` | Tail-allocated bitmap marking occupied entries |
    | `<KEYS>` | Tail-allocated buffer for keys |
    | `<VALUES>` | Tail-allocated buffer for values |

    The old two-word `bitmap` field has been removed.
- Introduce dedicated classes for the empty singleton instances. (`_EmptySetSingleton`, `_EmptyDictionarySingleton`.)
- Refactor deferred bridging. Add `_BridgingHashBuffer`, a standalone flat hash buffer for use in deferred bridging. Use it to improve memory use in cases where only one of Key or Value is bridged non-verbatim.
- Deferred bridging changes also eliminate the need for the `TypedNative*Storage` class, as well as for `_NativeSet`/`_NativeDictionary`’s support for non-hashable keys. Remove them.
- Rename storage classes as follows:

    | Old name | New name |
    |---|---|
    | `_RawNativeSetStorage` |  `_RawSetStorage` |
    | `_RawNativeDictionaryStorage` |  `_RawDictionaryStorage` |
    | `_TypedNativeSetStorage` |  *(removed)* |
    | `_TypedNativeDictionaryStorage` |  *(removed)* |
    | `_HashableTypedNativeSetStorage` |  `_SetStorage` |
    | `_HashableTypedNativeDictionaryStorage` |  `_DictionaryStorage` |
    | *(new)* | `_EmptySetSingleton` |
    | *(new)* | `_EmptyDictionarySingleton` |
    The new names make it obvious at runtime that the new ivar layout is in use; this is useful for tools like lldb.
- Clean up the implementations of Dictionary's various generalized accessors (`_modify`).

~~An lldb data formatter update will need to land before merging this PR.~~ The corresponding changes to lldb's data formatters are already in place: https://github.com/apple/swift-lldb/pull/927.

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke test

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke test

lorentey committed:
[stdlib] Add bitset constructs

Bitsets implement sorted sets over nonnegative integers up to a predetermined maximum value.

These are intended to replace _UnsafeBitMap. The latter will be removed once its usages are eliminated.

- _UnsafeBitset.Word is the underlying abstraction, implementing a bitset using a single UInt value.
- _UnsafeBitset is a view over a contiguous range of words.
- _Bitset is a COW value type implementing the same construct.

lorentey committed:
[stdlib] Hasher._Seed: Add internal typealias

This is intended to improve readability.

lorentey committed:
[stdlib] Add _HashTable implementing low-level hash table operations

_HashTable implements low-level hash table operations that are common between Set and Dictionary. It is a non-generic struct, dealing with maintaining hash table metadata rather than any actual elements.

_HashTable is intended as a mostly transparent abstraction. Most operations are inlinable, except for the ones related to the maximum load factor.

lorentey committed:
[stdlib] Remove _UnsafeBitMap

Use _UnsafeBitset instead.

lorentey committed:
[stdlib] _makeCollectionDescription: Make internal

lorentey committed:
[stdlib] _stdlib_NSObject_isEqual: Make internal

lorentey committed:
[stdlib] GroupInfo.json: Make a Hashing group

lorentey committed:
[stdlib] Set,  Dictionary: New storage classes for the empty singletons

This allows us to move the empty NSSet/NSDictionary overrides out of the root storage class; they don’t really belong there. More importantly, it makes empty sets/dictionaries super obvious to lldb and other runtime tools.

lorentey commented:
```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/test/SILOptimizer/no_traps_in_dict_loopup.swift:8:15: error: CHECK-NOT: excluded string found in input
// CHECK-NOT: llvm.trap
              ^
<stdin>:105:15: note: found here
declare void @llvm.trap() #1
              ^~~~~~~~~
```

Oops, `Dictionary.swift` accidentally uses the checked bitset lookup variants. I'm removing those in an upcoming cleanup. 

The lldb failures are to be expected.

lorentey committed:
[stdlib] Clean up Set and Dictionary

- Use _HashTable to unify low-level hashing operations across Set and Dictionary.
- Store the capacity directly inside the storage header. This allows the maximum load factor to be controlled by non-inlinable code.
- Introduce a dedicated class for the empty singleton.
- Add _BridgingHashBuffer, a standalone flat hash buffer for use in deferred bridging. Use it to eliminate the need to support non-hashable storage/wrapper variants and to improve memory use in cases where Key or Value are verbatim bridged.
- Eliminate the “TypedNative*Storage” class and _NativeSet/_NativeDictionary’s support for non-hashable keys.
- Rename storage classes as follows:

_RawNativeSetStorage ⟹ _RawSetStorage
_RawNativeDictionaryStorage ⟹ _RawDictionaryStorage
_TypedNativeSetStorage ⟹ (removed)
_TypedNativeDictionaryStorage ⟹ (removed)
_HashableTypedNativeSetStorage ⟹ _SetStorage
_HashableTypedNativeDictionaryStorage ⟹ _DictionaryStorage

The new names make it obvious which ivar layout is in use.

lorentey committed:
[test] Update tests for Set/Dictionary changes

Along with updating tests to adopt the new storage class names, the behavior of removeAll() has changed slightly.

If t stored an empty Set/Dictionary that was bridged from Objective-C, t.removeAlI() used to keep the original Objective-C storage intact. Now removeAll() replaces the old storage with the empty singleton, which makes a lot more sense to me.

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
The latest changes revert the experimental metadata layout changes -- the SIMD tricks they employed had excellent throughput, but their latency was too high for the kind of short chains that occur in normal hash tables. In the latest version `Set` and `Dictionary` use a simple flat bitmap data structure, like before.

Some of the SIMD-in-a-register tricks still survive in the `_HashTable` struct; they are proved useful for iterating over bitmaps, for example.

I split out some optimizations and bugfixes into separate PRs, to be submitted after this lands.

palimondo commented:
I haven't looked at the code, I'm just curious. Was the SIMD code x86 specific or did that work on ARM too?

lorentey commented:
The experimental metadata layout stored 8 extra bits of the hash value for each entry in the table. These were used to reduce the cost of collisions. The "SIMD" in this case was done entirely using general-purpose registers -- to process the metadata table, I treated a regular UInt64 as a vector of 8 bytes, so that I could search hash lookup chains by [looking at 8 entries at once][bithack].

[bithack]: https://github.com/lorentey/swift/blob/0da5e6425081d32ba555d97d17b4ba9f16065299/stdlib/public/core/HashTable.swift#L369-L383

This was a significant boost for long chains, but unfortunately, the average size of a lookup chain is between 1 and 2.5 -- so there was just not enough data for SIMD's advantages to kick in, and `Set.contains` got significantly slower.

palimondo commented:
Thanks for the explanation! What was the cause of slowdown in `contains` — just the change of data layout by adding the extra byte?

lorentey commented:
Yeah, processing 8x more metadata was much more costly than I expected. The entire point of that exercise was to allow us to increase the max load factor, but SIMD tricks weren't enough to paper over the problems with that, especially at table sizes < 2k. I also tried Robin hood hashing etc., but that had other drawbacks. 

The simple algorithm the stdlib's hashed collections use seems very difficult to beat on anything other than memory use.

lorentey commented:
I could reproduce the benchmark build failure; looks like the defaultValue subscript isn't exercised enough in the test suite.

palimondo commented:
Isn't it possible that your experimental design would outperform the simple algorithm on different architecture? I'm assuming you've tested on x86. How did it look on ARM? (Are we tracking benchmarks on that, too?)

lorentey committed:
[stdlib] Dictionary: Describe mutatingFind(_:)

lorentey committed:
[stdlib] Dictionary: prepareForMutation(at:) ⟹ ensureUniqueNative(preserving:)

Also describe it in a doc comment.

lorentey committed:
[stdlib] Dictionary: Don’t use the allKeys property of bridged indices directly

lorentey committed:
[stdlib] Dictionary: Attempt to work around ClosureSpecializer crash

rdar://problem/44612356

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please test

lorentey committed:
[stdlib] Dictionary: Fix compilation on non-ObjC platforms

palimondo commented:
Looks like the improvements from #16310 got lost in the shuffle...

lorentey commented:
The benchmark results are mostly as expected, except:

* Obviously I'll need to fix `DictionaryKeysContainsCocoa`.
* I'll look into why removals got slower -- an `@inlinable` may have been lost in the shuffle to get this ready. The new code uses a slightly different algorithm for removals that tends to move more entries around, but has better worst-case behavior. This did not affect my own benchmarks, but maybe I missed something.

Other than those, the results are in line with my expectations:

* Insertions got faster because `capacity` is now directly in the storage header. 
* It's faster to iterate over sparse sets/dictionaries because `_HashTable` skips over runs of empty entries without looping over each individual bucket.
* The code is more careful about moving (rather than copying, then releasing) elements when it can.
* Also, uniqueness checks are more consistently optimized for specific operations.

lorentey commented:
> Looks like the improvements from #16310 got lost in the shuffle...

That was intentional; I'll resubmit those once this lands. This PR is already ten times larger than what's manageable. :(

lorentey committed:
[stdlib] Fix performance regression in Dictionary.Keys.contains(_:)

Previously it called index(forKey:) to implement the test, but that method is O(n) for dictionaries bridged over from Cocoa.

lorentey committed:
[stdlib] Set, Dictionary: Make _HashTableDelegate implementations inlinable

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke benchmark

lorentey committed:
[stdlib] Set, Dictionary: Simplify variant processing

Don’t duplicate native paths — make the cocoa case conditionally unreachable instead.

lorentey committed:
[stdlib] Dictionary: Speed up Keys.subscript for bridged dictionaries

There is no need to fetch the value corresponding to the given key.
Also implement the same for Values.subscript, although the impact there is marginal.

lorentey committed:
[stdlib] Set, Dictionary: Speed up insert/remove when there’s nothing to do

Moving the uniqueness check before the first lookup was a bad idea. Revert it.

lorentey committed:
[stdlib] Add missing calls to cocoaPath()

This should reduce code size and perhaps trigger more optimizations.

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please smoke benchmark

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
It turns out I regressed the fast paths of insert and remove by moving the uniqueness check before the first lookup. (E.g. the code checked for uniqueness on insert even if the element was already in the set.) That's a silly mistake, but thankfully it's an easy fix.

palimondo commented:
What's causing the slight code size changes in benchmarks that don't test Set and Dictionary? Is it coming from tag sets used in `BenchmarkInfo`?

lorentey commented:
https://ci.swift.org/job/swift-PR-Linux/7603/

```
14:06:37 Unrecognized terminator leading to phi block

14:06:37 2.	While running pass #193971 SILFunctionTransform "SimplifyCFG" on SILFunction "@$sSh4hash4intoys6HasherVz_tF".
14:06:37  for 'hash(into:)' at /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/stdlib/public/core/Set.swift:467:10
14:06:37 #0 0x000000000458b6d4 PrintStackTraceSignalHandler(void*) (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0x458b6d4)
14:06:37 #1 0x00000000045894a0 llvm::sys::RunSignalHandlers() (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0x45894a0)
14:06:37 #2 0x000000000458b882 SignalHandler(int) (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0x458b882)
14:06:37 #3 0x00007fe165e99390 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x11390)
14:06:37 #4 0x00007fe1645d8428 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x35428)
14:06:37 #5 0x00007fe1645da02a abort (/lib/x86_64-linux-gnu/libc.so.6+0x3702a)
14:06:37 #6 0x000000000451e71d llvm::llvm_unreachable_internal(char const*, char const*, unsigned int) (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0x451e71d)
14:06:37 #7 0x0000000000f6fc35 getEdgeValuesForTerminator(swift::TermInst*, swift::SILBasicBlock*) (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0xf6fc35)
14:06:37 #8 0x0000000000f72510 llvm::SSAUpdaterImpl<swift::SILSSAUpdater>::CheckIfPHIMatches(swift::SILPHIArgument*) (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0xf72510)
14:06:37 #9 0x0000000000f71d0b llvm::SSAUpdaterImpl<swift::SILSSAUpdater>::FindExistingPHI(swift::SILBasicBlock*, llvm::SmallVectorImpl<llvm::SSAUpdaterImpl<swift::SILSSAUpdater>::BBInfo*>*) (/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift+0xf71d0b)
```

Well that's a new one.

lorentey commented on stdlib/public/core/Dictionary.swift:
```diff
@@ -3697,6 +3701,7 @@ extension Dictionary._Variant {
       return Dictionary.Iterator(_native: native.makeIterator())
 #if _runtime(_ObjC)
     case .cocoa(let cocoa):
+      cocoaPath()
```
This change allows the optimizer to eliminate the Cocoa path for Element types that aren't bridged verbatim. My guess is that this triggers the SILFunctionTransform crash -- I'll verify tomorrow.

lorentey replied:
Ah, except the crash occurs on Linux. Hm.

lorentey committed:
[stdlib] Set.Iterator.next(): Work around compiler crash in non-ObjC builds

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please test linux platform

lorentey committed:
[stdlib] Temporary workaround for issue fixed in #19444

lorentey commented:
Let's try testing with the following PR. I think there will still be a couple small test failures, but the compiler crashes should be gone now.

https://github.com/apple/swift-lldb/pull/927
@swift-ci please test

lorentey commented:
apple/swift-lldb#927
@swift-ci please test

lorentey commented:
https://github.com/apple/swift-lldb/pull/927
@swift-ci please test

lorentey committed:
[stdlib] Dictionary.removeValue(forKey:): Fix typo.

lorentey committed:
[stdlib] Bitset: Eliminate a trap that wasn’t optimized away

Signed/unsigned integer conversions check for unrepresentable values; this wasn’t recognized as impossible, so a trap got compiled into the Dictionary lookup path.

Note to self: next time just use bitwise operations.

lorentey commented:
The latest two commits fix all issues I've seen locally. (The Linux build above failed in SIL/parse_stdlib.sil, which I haven't seen yet.)

Test with the following PR:
https://github.com/apple/swift-lldb/pull/927
@swift-ci please test

lorentey commented:
Test with the following PR:
https://github.com/apple/swift-lldb/pull/927
@swift-ci please test

lorentey commented:
Test with the following PR:
https://github.com/apple/swift-lldb/pull/927
@swift-ci please test

lorentey commented:
@swift-ci Please test compiler performance

lorentey commented:
@swift-ci Please test compiler performance

lorentey commented:
@swift-ci Please test compiler performance

lorentey commented:
Almost there! Now we just need to fix 32-bit platforms and SIL/parse_stdlib.sil.

lorentey committed:
[test] Fix flaky test

lorentey committed:
[stdlib] Remove _Bitset type for now

It is currently unused, but it’s causing issues with the SIL/parse-stlib.sil test.

lorentey commented:
@swift-ci please text linux platform

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please test linux platform

lorentey committed:
[test] Remove _Bitset tests

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please test linux platform

lorentey commented:
@swift-ci please test linux platform

lorentey committed:
[runtime] Fix layout of empty Set/Dictionary singletons

lorentey committed:
[stdlib] Discourage CoW copying / rehashing from getting inlined

lorentey committed:
[stdlib] _BridgingHashBuffer: Fix memory management

- Have the hash buffer include a reference to the original hash storage instance, along with a copy of its _HashTable, so that its lifetime can be independent from the deferred bridging object.
- Convert _BridgingHashBuffer to a ManagedBuffer so that we can easily put reference-counted properties in its header.

lorentey committed:
[stdlib] Set, Dictionary: Sanitize @objc-ness for NSEnumerator members

lorentey commented:
@swift-ci please test macOS platform

lorentey commented:
@swift-ci please test macOS platform

lorentey commented:
@swift-ci please test macOS platform

lorentey committed:
[stdlib] Dictionary: Remove workaround for rdar://problem/44612356

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
@swift-ci please test

lorentey commented:
Rebased to latest master. Latest round of commits fix all known regressions. 

This is finally getting ready to land.

lorentey commented:
I think it's time to get this one out of the way! ✨

Optimizations I pulled back from the original will come back as followup PRs soon.
